lib/MailNotifications.php

```diff
@@ -0,0 +1,330 @@
+ *
+ */
+
+namespace OCA\Notifications;
+
+use OCP\AppFramework\Utility\ITimeFactory;
+use OCP\Defaults;
+use OCP\IConfig;
+use OCP\IDateTimeFormatter;
+use OCP\IURLGenerator;
+use OCP\IUserManager;
+use OCP\L10N\IFactory;
+use OCP\Mail\IMailer;
+use OCP\Mail\IMessage;
+use OCP\Notification\IManager;
+use OCP\Notification\INotification;
+use OCP\Notification\IAction;
+use OCP\Util;
+use Psr\Log\LoggerInterface;
+
+class MailNotifications {
+
+	/** @var IConfig */
+	private $config;
+
+	/** @var IManager */
+	private $manager;
+
+	/** @var Handler */
+	protected $handler;
+
+	/** @var IUserManager */
+	private $userManager;
+
+	/** @var LoggerInterface */
+	private $logger;
+
+	/** @var IMailer */
+	private $mailer;
+
+	/** @var IURLGenerator */
+	private $urlGenerator;
+
+	/** @var Defaults */
+	private $defaults;
+
+	/** @var IFactory */
+	private $l10nFactory;
+
+	/** @var IDateTimeFormatter */
+	private $dateFormatter;
+
+	/** @var ITimeFactory */
+	protected $timeFactory;
+
+	public const DEFAULT_BATCH_TIME = 3600 * 24;
+
+	public function __construct(
+		IConfig $config,
+		IManager $manager,
+		Handler $handler,
+		IUserManager $userManager,
+		LoggerInterface $logger,
+		IMailer $mailer,
+		IURLGenerator $urlGenerator,
+		Defaults $defaults,
+		IFactory $l10nFactory,
+		IDateTimeFormatter $dateTimeFormatter,
+		ITimeFactory $timeFactory
+	) {
+		$this->config = $config;
+		$this->manager = $manager;
+		$this->handler = $handler;
+		$this->userManager = $userManager;
+		$this->logger = $logger;
+		$this->mailer = $mailer;
+		$this->urlGenerator = $urlGenerator;
+		$this->defaults = $defaults;
+		$this->l10nFactory = $l10nFactory;
+		$this->dateFormatter = $dateTimeFormatter;
+		$this->timeFactory = $timeFactory;
+	}
+
+	/**
+	 * Send all due notification emails.
+	 */
+	public function sendEmails(): void {
+		// Get all users who enabled notification emails.
+		$users = $this->config->getUsersForUserValue('notifications', 'notifications_email_enabled', '1');
+
+		if (count($users) == 0) {
+			return;
+		}
+
+		// Batch-read settings that will be used to figure out who needs notifications sent out
+		$userBatchTimes = $this->config->getUserValueForUsers('notifications', 'notify_setting_batchtime', $users);
+		$userLastSendIds = $this->config->getUserValueForUsers('notifications', 'mail_last_send_id', $users);
+		$userLastSendTimes = $this->config->getUserValueForUsers('notifications', 'mail_last_send_time', $users);
+
+		$now = $this->timeFactory->getTime();
+
+		foreach ($users as $user) {
+			// Get the settings for this particular user, then check if we have notifications to email them
+			$batchTime = (array_key_exists($user, $userBatchTimes)) ? (int) $userBatchTimes[$user] : self::DEFAULT_BATCH_TIME;
+			$lastSendId = (array_key_exists($user, $userLastSendIds)) ? (int) $userLastSendIds[$user] : -1;
+			$lastSendTime = (array_key_exists($user, $userLastSendTimes)) ? (int) $userLastSendTimes[$user] : -1;
```

Comment:
Only a suggestion to make more clean using null coalescing operator:

```suggestion
			$batchTime = (int) ($userBatchTimes[$user] ?? self::DEFAULT_BATCH_TIME);
			$lastSendId = (int) ($userLastSendIds[$user] ?? -1);
			$lastSendTime = (int) ($userLastSendTimes[$user] ?? -1);
```